cmd/admin/auth.go

```diff
@@ -102,8 +103,45 @@ func handlerAuthCheck(h http.Handler) http.Handler {
 				samlMiddleware.RequireAccount(h).ServeHTTP(w, r)
 			}
 		case settings.AuthHeaders:
-			// Access always granted
-			h.ServeHTTP(w, r)
+			username := r.Header.Get(headersConfig.TrustedPrefix + headersConfig.UserName)
+			groups := strings.Split(r.Header.Get(headersConfig.TrustedPrefix+headersConfig.Groups), ",")
+
+			if username == "" {
+				log.Printf("A username is required to use this system.")
+				w.WriteHeader(http.StatusBadRequest)
+				return
+			}
+
+			s := make(contextValue)
+			s["user"] = username
+
+			for _, group := range groups {
+				if group == headersConfig.AdminGroup {
+					s["level"] = adminLevel
+					// We can break because there is no greater permission level
+					break
+				} else if group == headersConfig.UserGroup {
+					s["level"] = userLevel
+					// We can't break because we might still find a higher permission level
+				}
+			}
+
+			// This user didn't present a group that has permission to use the service
+			if _, ok := s["level"]; !ok {
+				w.WriteHeader(http.StatusForbidden)
+				return
+			}
+
+			err := adminUsers.UpdateMetadata(r.RemoteAddr, r.Header.Get("User-Agent"), username)
```

Comment:
Since this authentication method does not require to pre-create users, better follow these steps:
1 - Check if user already exists
2 - If not, create user
3 - Update metadata
I would replace this block (135-138) with the following code:
```
if !adminUsers.Exists(username) {
  log.Printf("user not found, creating new user: %s", username)
  email := r.Header.Get(headersConfig.TrustedPrefix + headersConfig.Email)
  fullname := r.Header.Get(headersConfig.TrustedPrefix + headersConfig.DisplayName)
  newUser, err := adminUsers.New(username, "", email, fullname, (s["level"] == adminLevel))
  if err != nil {
    log.Printf("error creating user %s: %v", username, err)
    http.Redirect(w, r, forbiddenPath, http.StatusFound)
    return
  }
} else {
  if err := adminUsers.UpdateMetadata(r.RemoteAddr, r.Header.Get("User-Agent"), username); err != nil {
    log.Printf("error updating metadata for user %s: %v", username, err)
  }
}
```